src/test/java/com/jcabi/s3/AwsBucketTest.java

```diff
@@ -106,4 +115,119 @@ public void existsThrowsIOException() throws IOException {
+        final Region region = Mockito.mock(Region.class);
+        new RegionExpectations(region)
+            .expectResponse(firstItem, null, secondItem)
+            .expectResponse(secondItem, secondItem, null)
+            .apply(bucketName, prefix);
+        final Bucket bucket = new AwsBucket(region, bucketName);
+        final Iterator<String> actual = bucket.list(prefix).iterator();
+        MatcherAssert.assertThat(actual.hasNext(), Matchers.equalTo(true));
+        MatcherAssert.assertThat(firstItem, Matchers.equalTo(actual.next()));
+        MatcherAssert.assertThat(actual.hasNext(), Matchers.equalTo(true));
+        MatcherAssert.assertThat(secondItem, Matchers.equalTo(actual.next()));
+        MatcherAssert.assertThat(actual.hasNext(), Matchers.equalTo(false));
+    }
+
+    private static class RegionExpectations {
+
+        /**
+         * Mocked s3 service.
+         */
+        private final transient AmazonS3 aws;
+
+        /**
+         * Responses.
+         */
+        private final transient List<ObjectListing> responses =
+            new ArrayList<ObjectListing>(0);
+
+        /**
+         * Expected markers.
+         */
+        private final transient List<String> markers = new ArrayList<String>(0);
+
+        /**
+         * Constructs region expectations.
+         * @param region Mocked region
+         */
+        public RegionExpectations(final Region region) {
+            this.aws = Mockito.mock(AmazonS3.class);
+            Mockito.when(region.aws()).thenReturn(this.aws);
+        }
+
+        /**
+         * Expect request with start marker and provide in response single item,
+         * notify that marker is a next marker to request.
+         * @param item Item to respond with
+         * @param start Start marker to expect
+         * @param marker Next marker
+         * @return This instance
+         */
+        public RegionExpectations expectResponse(final String item,
+            final String start, final String marker) {
+            final ObjectListing response = Mockito.mock(ObjectListing.class);
+            Mockito.when(response.getNextMarker()).thenReturn(marker);
+            final List<S3ObjectSummary> summaries =
+                new ArrayList<S3ObjectSummary>(1);
+            final S3ObjectSummary summary = new S3ObjectSummary();
+            summary.setKey(item);
+            summaries.add(summary);
+            Mockito.when(response.getObjectSummaries())
+                .thenReturn(summaries);
+            this.responses.add(response);
+            this.markers.add(start);
+            return this;
+        }
+
+        /**
+         * Apply expectations.
+         * @param bucket Bucket name
+         * @param prefix Request prefix
+         */
+        public void apply(final String bucket, final String prefix) {
+            Mockito.when(
+                this.aws.listObjects(Mockito.any(ListObjectsRequest.class))
+            ).thenAnswer(
```

Comment:
you can use `Mockito.when().thenAnswer().thenAnswer().thenAnswer()...` I think this should help. Now the code looks very difficult to read. I tried three times.